pdu/UDH.go

```diff
@@ -15,27 +17,35 @@ import (
 type UDH []InfoElement
 
 // UDHL return the length (number of octet) of the encoded UDH itself
-func (u UDH) UDHL() (l int) {
-	for i := range u {
-		l += len(u[i].Data)
+func (u *UDH) UDHL() int {
+	if u == nil {
+		return 0
 	}
-	return l
+	length := 1
```

Comment:
UDH is not only for Concatenated SMS. Please recheck the spec. cc @bachhh for reviewing this

Reply:
reference *UDH* memory layout.

```
length (1 byte)
[data] ([length - 1] byte)
```

info element:
```
length (1 byte)
id (1 byte)
[data] ([length - 2] byte)
```

This algorithm is no problem

Reply:
@bachhh wrong implementation `InfoElement.Data`

https://github.com/linxGnu/gosmpp/blob/cc603b5cc822dcd18a9d1e72b65158f773be5db5/pdu/UDH.go#L72

The correct logic should be

```go
src[ri] // length indicator
src[ri + 1] // id
src[ri + 2: ri + 2 + iedl] // data area

ie.Data = append([]byte(nil), src[ri+2:ri+2+iedl]...)
```

Reply:
Thanks @septs , I was thinking ie.Data should wraps up all data for a single IE including ID & length. But doing it this way is way less confusing.